internal/ota/ota.go

```diff
@@ -0,0 +1,432 @@
+	if err != nil {
+		return nil, fmt.Errorf("error decoding response: %w", err)
+	}
+
+	traceLogger().
+		Msg("fetchUpdateMetadata: completed")
+
+	return metadata, nil
+}
+
+func (s *State) triggerStateUpdate() {
+	s.onStateUpdate(s.ToRPCState())
+}
+
+func (s *State) triggerComponentUpdateState(component string, update *componentUpdateStatus) {
+	s.componentUpdateStatuses[component] = *update
+	s.triggerStateUpdate()
+}
+
+// TryUpdate tries to update the given components
+// if the update is already in progress, it returns an error
+func (s *State) TryUpdate(ctx context.Context, params UpdateParams) error {
+	locked := s.mu.TryLock()
+	if !locked {
+		return fmt.Errorf("update already in progress")
+	}
+
+	return s.doUpdate(ctx, params)
+}
+
+// before calling doUpdate, the caller must have locked the mutex
+// otherwise a runtime error will occur
+func (s *State) doUpdate(ctx context.Context, params UpdateParams) error {
+	defer s.mu.Unlock()
+
+	scopedLogger := s.l.With().
+		Interface("params", params).
+		Logger()
+
+	scopedLogger.Info().Msg("checking for updates")
+	if s.updating {
+		return fmt.Errorf("update already in progress")
+	}
+
+	s.updating = true
+	s.triggerStateUpdate()
+
+	if len(params.Components) == 0 {
+		params.Components = defaultComponents
+	}
+
+	_, shouldUpdateApp := params.Components["app"]
+	_, shouldUpdateSystem := params.Components["system"]
+
+	if !shouldUpdateApp && !shouldUpdateSystem {
+		return s.componentUpdateError(
+			"Update aborted: no components were specified to update. Requested components: ",
+			fmt.Errorf("%v", params.Components),
+			&scopedLogger,
+		)
+	}
+
+	appUpdate, systemUpdate, err := s.getUpdateStatus(ctx, params)
+	if err != nil {
+		return s.componentUpdateError("Error checking for updates", err, &scopedLogger)
+	}
+
+	s.metadataFetchedAt = time.Now()
+	s.triggerStateUpdate()
+
+	if shouldUpdateApp && appUpdate.available {
+		appUpdate.pending = true
+		s.updating = true
+		s.triggerComponentUpdateState("app", appUpdate)
+	}
+
+	if shouldUpdateSystem && systemUpdate.available {
+		systemUpdate.pending = true
+		s.updating = true
+		s.triggerComponentUpdateState("system", systemUpdate)
+	}
+
+	scopedLogger.Trace().Bool("pending", appUpdate.pending).Msg("Checking for app update")
+
+	if appUpdate.pending {
+		scopedLogger.Info().
+			Str("url", appUpdate.url).
+			Str("hash", appUpdate.hash).
+			Msg("App update available")
+
+		if err := s.updateApp(ctx, appUpdate); err != nil {
+			return s.componentUpdateError("Error updating app", err, &scopedLogger)
+		}
+	} else {
+		scopedLogger.Info().Msg("App is up to date")
+	}
+
+	scopedLogger.Trace().Bool("pending", systemUpdate.pending).Msg("Checking for system update")
+
+	if systemUpdate.pending {
+		if err := s.updateSystem(ctx, systemUpdate); err != nil {
+			return s.componentUpdateError("Error updating system", err, &scopedLogger)
+		}
+	} else {
+		scopedLogger.Info().Msg("System is up to date")
+	}
+
+	s.updating = false
```

Comment:
Should we just wrap this in a `defer` up higher in the function to ensure we always clear the flag?